README.md
nagios-checks
=============

![Build](https://github.com/benprew/nagios-checks/workflows/Build/badge.svg)

Various Nagios check scripts.

check_haproxy
-------------

Checks haproxy stats and reports errors if any of the servers for a proxy are down.

    Usage: check_haproxy.rb [options]

    Specific options:
        -u, --url URL                    Statistics URL to check (e.g. http://demo.1wt.eu/)
        -p, --proxies [PROXIES]          Only check these proxies (eg proxy1,proxy2,proxylive)
        -U, --user [USER]                Basic auth user to login as
        -P, --password [PASSWORD]        Basic auth password
        -w, --warning [WARNING]          Threshold for number of sessions as a percentage of the limit
        -c, --critical [CRITICAL]        Threshold for number of sessions as a percentage of the limit
        -h, --help                       Display this screen

Example: ```check_haproxy.rb -u "http://demo.1wt.eu/" -w 80 -c 95```


check_solr_slave
----------------

    Usage: check_solr_slave [options]
    -H, --hostname [hostname]        Host to connect to [localhost]
    -p, --port [port]                Port to connect to [8983]
    -w, --warn [minutes]             Threshold for warning [15]
    -c, --crit [minutes]             Threshold for critical [30]
    -h, --help                       Display this screen

License
-------

GPL https://www.gnu.org/licenses/gpl.html


CHANGELOG.md
# check_haproxy.rb CHANGELOG

## v3.2.0
* Add cookie argument to check_haproxy. Fixes #15
* Fix bug with checking header variable
* Enable openssl as needed.  fixes #9
* add --http-error-critical
* parameters for SSL and insecure certificate
* fix redirect for ruby 1.8 and change return codes
* try first redirect which is mainly to https
* allow https with invalid certificate on Ruby 1.8+
* Adding some error management
* check_haproxy: collect perfdata from frontends, backends and servers.
## v3.1.0
* Ignore backends without checks
* Better plugin output; added thresholds for number of sessions
* Add a little bit more output by default, but disable the "UP" spam unless you specify --debug (option was previously unused).
* print all down proxies on a single line for nagios status
## v3.0.0
* Fix bug when incorrectly reporting status 'UP 1/3' as warning
* add debug flag, update usage and make script work with ruby 1.8+
* always display output and list down proxies first
* Use method not hash key access
* add license details
## v1.0.0
* Initial version


test/test_check_haproxy.sh
#!/bin/bash

set -e

echo -n "testing nothing smokes..."
for f in test/haproxy/*; do
    ./check_haproxy.rb -u "$f" |head -n1
done > /tmp/output.txt

cat <<EOF > /tmp/expected.txt
HAPROXY CRIT: www www DOWN active 	sess=0/20(0%) smax=14; git www DOWN active 	sess=0/2(0%) smax=2
HAPROXY OK: 58 proxies found
HAPROXY CRIT: osbs-backend osbs-master01 DOWN active 	sess=0/-1(0%) smax=0; osbs-backend BACKEND DOWN 	sess=0/500(0%) smax=0
HAPROXY OK: 13 proxies found
HAPROXY CRIT: ubuntuusers-tt srv10 DOWN active 	sess=0/512(0%) smax=0; ubuntuusers-tt BACKEND DOWN 	sess=0/820(0%) smax=1; ubuntuusers-static kanu DOWN active 	sess=0/512(0%) smax=0; ubuntuusers-media kanu DOWN active 	sess=0/512(0%) smax=0; ubuntuusers-tour kanu DOWN active 	sess=0/512(0%) smax=0; ubuntuusers-legacy kanu DOWN active 	sess=0/512(0%) smax=0
EOF

diff -u /tmp/output.txt /tmp/expected.txt
echo "OK"

echo -n "testing warn limit..."
./check_haproxy.rb -u "test/haproxy/fedoraproject_org.csv" -w 2 -p fedmsg-raw-zmq-outbound-backend |grep 'WARN.*too many sessions' > /dev/null
echo "OK"

echo -n "testing crit limit..."
./check_haproxy.rb -u "test/haproxy/fedoraproject_org.csv" -w 1 -c2 -p fedmsg-raw-zmq-outbound-backend |grep 'CRIT.*too many sessions' > /dev/null
echo "OK"

echo "SUCCESS! ALL TESTS PASSED"


check_solr_slave.rb
#!/usr/bin/env ruby

require 'rexml/document'
require 'uri'
require 'net/http'
require 'optparse'
require 'date'

# Set default options
options = {}
options[:hostname] = "localhost"
options[:port] = "8983"
options[:warn] = "15"
options[:crit] = "30"

# Parse command line options
opts = OptionParser.new
opts.on('-H', '--hostname [hostname]', 'Host to connect to [localhost]') do |hostname|
  options[:hostname] = hostname
end

opts.on('-p', '--port [port]', 'Port to connect to [8983]') do |port|
  options[:port] = port
end

opts.on('-w', '--warn [minutes]', 'Threshold for warning [15]') do |warn|
  options[:warn] = warn
end

opts.on('-c', '--crit [minutes]', 'Threshold for critical [30]') do |crit|
  options[:crit] = crit
end

opts.on( '-h', '--help', 'Display this screen' ) do
  puts opts
  exit 3
end
opts.parse!

def get_status_from_uri(uri)
  begin
    res = Net::HTTP.get(uri)
  rescue SocketError => e
    puts "CRITICAL - Unable to contact #{uri}: HTTP #{e}"
    exit 2
  end

  REXML::Document.new(res)
end

def index_version(rexml_doc)
  REXML::XPath.first(rexml_doc, '*//long[@name="replicatableIndexVersion"]').text.to_i
end

solr_slave_status = get_status_from_uri(URI("http://#{options[:hostname]}:#{options[:port]}/solr/replication?command=details"))
solr_master_status = get_status_from_uri(URI(REXML::XPath.first(solr_slave_status, '*//str[@name="masterUrl"]').text + "?command=details"))
replication_time = DateTime.parse(REXML::XPath.first(solr_slave_status, '*//str[@name="indexReplicatedAt"]').text)

# First check to see if we have matching index numbers
# If they don't match, how long since we have succeeded?
current_time = DateTime.now
if index_version(solr_master_status) == index_version(solr_slave_status)
  puts "OK - Slave is current: #{replication_time}"
  exit 0
elsif replication_time > current_time - options[:warn].to_i / 1440.0
  puts "OK - Slave is recent: #{replication_time}"
  exit 0
elsif replication_time < current_time - options[:crit].to_i / 1440.0
  puts "CRITICAL - Slave more than #{options[:crit]} minutes old: #{replication_time}"
  exit 2
elsif replication_time < current_time - options[:warn].to_i / 1440.0
  puts "WARNING - Slave more than #{options[:warn]} minutes old: #{replication_time}"
  exit 1
else
  puts "UNKNOWN - Unable to determine replication status"
  exit 3
end

=begin
Copyright (C) 2013 Ben Prew

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
=end


check_haproxy.rb
#!/usr/bin/ruby

require 'optparse'
require 'open-uri'
require 'ostruct'
require 'csv'

VERSION = :'3.2.0'

OK = 0
WARNING = 1
CRITICAL = 2
UNKNOWN = 3


status = ['OK', 'WARN', 'CRIT', 'UNKN']

@proxies = []
@errors = []
@perfdata = []
options = OpenStruct.new
options.proxies = []

op = OptionParser.new do |opts|
  opts.banner = 'Usage: check_haproxy.rb [options]'

  opts.separator ''
  opts.separator 'Specific options:'

  # Required arguments
  opts.on("-u", "--url URL", "Statistics URL to check (eg. http://demo.1wt.eu/)") do |v|
    options.url = v
    options.url += "/;csv" unless options.url =~ /csv$/
  end

  # Optional Arguments
  opts.on("-p", "--proxies [PROXIES]", "Only check these proxies (eg. proxy1,proxy2,proxylive)") do |v|
    options.proxies = v.split(/,/)
  end

  opts.on("-U", "--user [USER]", "Basic auth user to login as") do |v|
    options.user = v
  end

  opts.on("-P", "--password [PASSWORD]", "Basic auth password") do |v|
    options.password = v
  end

  opts.on("-w", "--warning [WARNING]", "Pct of active sessions (eg 85, 90)") do |v|
    options.warning = v
  end

  opts.on("-c", "--critical [CRITICAL]", "Pct of active sessions (eg 90, 95)") do |v|
    options.critical = v
  end

  opts.on('-k', '--insecure', 'Allow insecure TLS/SSL connections') do
    options.insecure_ssl = true
  end

  opts.on('--http-error-critical', 'Throw critical when connection to HAProxy is refused or returns error code') do
    options.http_error_critical = true
  end

  opts.on('--cookie [COOKIE]', 'Login/Session cookie') do |v|
    options.cookie = v
  end

  opts.on('-h', '--help', 'Display this screen') do
    puts opts
    exit 3
  end

  opts.on('--version', 'Show version and exit') do
    puts "check_haproxy.rb v#{VERSION}"
    exit 3
  end
end

op.parse!

unless options.url
  puts 'ERROR: URL is required'
  puts op
  exit UNKNOWN
end

if options.warning && !options.warning.to_i.between?(0, 100)
  puts 'ERROR: warning must be between 0 and 100'
  puts op
  exit UNKNOWN
end

if options.critical && !options.critical.to_i.between?(0, 100)
  puts 'ERROR: critical must be between 0 and 100'
  puts op
  exit UNKNOWN
end

if options.warning && options.critical && options.warning.to_i > options.critical.to_i
  puts 'ERROR: warning must be below critical'
  puts op
  exit UNKNOWN
end

def open_options(options)
  open_opts = {
    :http_basic_authentication => [options.user, options.password]
  }
  open_opts['Cookie'] = options.cookie if options.cookie

  # allows https with invalid certificate on ruby 1.9
  # src: http://snippets.aktagon.com/snippets/370-hack-for-using-openuri-with-ssl
  if options.insecure_ssl && RUBY_VERSION =~ /1\.9/
    open_opts[:ssl_verify_mode] = OpenSSL::SSL::VERIFY_NONE
  end

  return open_opts
end

def haproxy_response(options)
  tries = 2

  if options.url =~ /https/
    require 'openssl'
    # allows https with invalid certificate on ruby 1.8
    # src: http://snippets.aktagon.com/snippets/370-hack-for-using-openuri-with-ssl
    if options.insecure_ssl && RUBY_VERSION =~ /1\.8/
      OpenSSL::SSL.const_set :VERIFY_PEER, OpenSSL::SSL::VERIFY_NONE
    end
  end

  begin
    return open(options.url, open_options(options))
  rescue OpenURI::HTTPError => e
    puts "ERROR: #{e.message}"
    options.http_error_critical ? exit(CRITICAL) : exit(UNKNOWN)
  rescue Errno::ECONNREFUSED => e
    puts "ERROR: #{e.message}"
    options.http_error_critical ? exit(CRITICAL) : exit(UNKNOWN)
  rescue RuntimeError => e
    if e.message =~ /redirection forbidden/
      options.url = e.message.gsub(/.*-> (.*)/, '\1')  # extract redirect URL
      retry if (tries -= 1) > 0
      raise
    else
      exit UNKNOWN
    end
  end
end

def exit_code(code=0)
  @exit_code ||= OK
  @exit_code = code if code > @exit_code
  @exit_code
end

header = nil

haproxy_response(options).each do |line|
  if line =~ /^# /
    header = line[2..-1].split(',')
    next
  elsif !header
    puts "ERROR: CSV header is missing"
    exit UNKNOWN
  end

  row = header.zip(CSV.parse_line(line)).reduce({}) { |hash, val| hash.merge({val[0] => val[1]}) }

  next unless options.proxies.empty? || options.proxies.include?(row['pxname'])
  next if ['statistics', 'admin_stats', 'stats'].include? row['pxname']
  next if row['status'] == 'no check'

  role = row['act'].to_i > 0 ? 'active ' : (row['bck'].to_i > 0 ? 'backup ' : '')
  if row['slim'].to_i == 0
    session_percent_usage = 0
  else
    session_percent_usage = row['scur'].to_i * 100 / row['slim'].to_i
  end

  proxy_name = sprintf("%s %s %s %s", row['pxname'], row['svname'], row['status'], role)
  message = sprintf("%s\tsess=%s/%s(%d%%) smax=%s",
                    proxy_name,
                    row['scur'],
                    row['slim'] || '-1',
                    session_percent_usage,
                    row['smax'])
  @proxies << message

  if row['status'].include?('DOWN')
    @errors << message
    exit_code CRITICAL
  end

  if options.critical && session_percent_usage >= options.critical.to_i
    @errors << sprintf("%s - too many sessions %s/%s(%d%%)", proxy_name, row['scur'], row['slim'], session_percent_usage)
    exit_code CRITICAL
  elsif options.warning && session_percent_usage >= options.warning.to_i
    @errors << sprintf("%s - too many sessions %s/%s(%d%%)", proxy_name, row['scur'], row['slim'], session_percent_usage)
    exit_code WARNING
  end
end

if @errors.length == 0
  @errors << sprintf("%d proxies found", @proxies.length)
end

if @proxies.length == 0
  @errors << "No proxies listed as up or down"
  exit_code UNKNOWN
end

puts "HAPROXY " + status[exit_code] + ": " + @errors.join('; ')
puts @proxies.sort

exit exit_code

=begin
Copyright (C) 2013 Ben Prew
Copyright (C) 2013 Mark Ruys, Peercode <mark.ruys@peercode.nl>
Copyright (C) 2015 Hector Sanjuan. Nugg.ad <hector.sanjuan@nugg.ad>
Copyright (C) 2015 Roger Torrentsgeneros <roger.torrentsgeneros@softonic.com>
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
=end
